### DISCUSSION TITLE
TF-IDF behaviour with stop words :-)

### DISCUSSION THREAD

**Discussion by AnneEstoppey** [MaartenGr/BERTopic#824] [Q&A]
Hello

I am trying to do topic modelling with BERTopic, with a collection of about 2000 texts (public data).

For reference I have already done topic modelling with gensim, using spaCy for language model and pyLDAvis for visualization of the results.
I got some results that made sense with 6 distinct topics. 

My goal here is to experiment with BERTopic, which I found quite interesting to start with! 
Therefore I am taking the same collection of texts.
My texts have no HTML tags, and are clear text to read by humans (punctuation is still there). Language is English.

I checked the bertopic colab notebooks from the documentation, and these were very helpful.

I have tried different scenarios, but I am not getting even close to the result that I had with gensim.
I am of course expecting differences, but at this stage I don't even have some ok ish topics. It's a "big salad".

My understanding was that if one uses TF-IDF as custom embedder, one would not need to remove stop words, since 'ubiquitous' words will not be taken into account while creating the topics. But that seems to be very incorrect, see my screenshot below: I am expecting stop words 'the', 'to', 'in', 'and', 'was', 'to' for example, not to be in the top topics and only in the '-1' topic. But they are still there in all topics!

![image](https://user-images.githubusercontent.com/35219455/200310716-c60209e6-50ec-4334-a413-8917bb155388.png)

Here is an example of one document (text):
![image](https://user-images.githubusercontent.com/35219455/200312081-a2fb862a-7c8d-4dc6-962d-77e1a612be5f.png)

Does that mean that one has to remove stop words from the documents beforehand? Even when using TF-IDF?

Thanks a lot in advance for hints and tips,
Cheers
Anne

**Comment by MaartenGr** [comment]
The thing with TF-IDF is that it needs a sufficient amount of data in order to differentiate stop words. In practice, you can quickly achieve that if you have a couple of hundred documents. In BERTopic, however, we have aggregates of documents instead of localized documents. Moreover, we are not considering TF-IDF but c-TF-IDF which approaches the task a bit differently than the original. 

This means that when you only have 2000 documents, then it is likely that those stop words are not properly excluded from the representations. Fortunately, that can be fixed quite easily by either removing stop words explicitly through the `CountVectorizer` or by focusing on the `ClassTfidfTransformer` instead which has a parameter for reducing the appearance of frequent words: 

```python
from bertopic import BERTopic
from sklearn.feature_extraction.text import CountVectorizer
from bertopic.vectorizers import ClassTfidfTransformer

# Using the CountVectorizer to reduce stop words
vectorizer_model = CountVectorizer(stop_words="english")
topic_model = BERTopic(vectorizer_model=vectorizer_model)

# Using the ClassTfidfTransformer to reduce stop words
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)
topic_model = BERTopic(ctfidf_model=ctfidf_model)
```

Do note that we do not have to remove stopwords before using BERTopic. The pipeline of BERTopic is generally as follows:

embeddings -> dimensionality reduction -> clustering -> topic extraction

If we remove stop words and use a large language model then that typically will result in poor clusters as you remove contextual information. In other words, removing stop words before creating the embeddings will have little influence on the number of stop words that you see in the output. This is because when we perform the topic extraction with c-TF-IDF, we still consider stop words as that process is separate from the embedding step. You can also find an overview of the pipeline [here](https://maartengr.github.io/BERTopic/algorithm/algorithm.html).

**Reply by AnneEstoppey** [reply]
Thanks a lot Maarten! :-) 

Yes in the meantime I read about c-TF-IDF versus TF-IDF, and thought that c-TF-IDF would be more appropriate, so thanks for elaborating, that makes complete sense :-)

I will test your above suggestions further, may come back here with results, so that people can see the difference as well :-)
